# MADJI NDOGO WATER  CRISIS

In [15]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook.
%reload_ext sql

displaylimit: Value None will be treated as 0 (no limit)

In [22]:
# Establish a connection to the local database using the '%sql' magic command
%sql mysql+pymysql://root:ShUfFlEs%401999@localhost:3306/md_water_services

Connecting and switching to connection 'mysql+pymysql://root:***@localhost:3306/md_water_services'

In [4]:
# Removee limit of 10 from queries 
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

# DATA CLEANING

Ok, bring up the employee table. It has info on all of our workers, but note that the email addresses have not been added. We will have to send
our workers reports and figures, so let's update it. Luckily the emails for our department are easy: first_name.last_name@ndogowater.gov.

In [32]:
%%sql 
# Call up the employee table

SELECT 
    *
FROM  
    md_water_services.employee
LIMIT 15;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

15 rows affected.

+----------------------+----------------+---------------+-------+----------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  |  phone_number | email |          address           | province_name | town_name |    position    |
+----------------------+----------------+---------------+-------+----------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287  |  None |  36 Pwani Mchangani Road   |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786  |  None |   129 Ziwa La Kioo Road    |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041  |  None |   18 Mlima Tazama Avenue   |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900  |  None |     100 Mogadishu Road     |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640  |  None |      1 Savanna Street      |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111  |  None |  26 Bahari Ya Faraja Road  |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631  |  None |    104 Kenyatta Street     |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240  |  None | 145 Sungura Amanpour Road  |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202  |  None |      117 Kampala Road      |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739  |  None | 33 Angélique Kidjo Avenue  |     Amanzi    |   Dahabu  | Field Surveyor |
|          12          |  Gamba Shani   | +99438283153  |  None |   38 Mlima Tazama Avenue   |    Hawassa    |   Rural   | Field Surveyor |
|          14          |  Harith Nyota  | +99917230124  |  None |      42 Malabo Drive       |    Kilimani   |   Rural   | Field Surveyor |
|          16          |  Isoke Amani   | +99313603182  |  None |       55 Fennec Way        |     Sokoto    |   Rural   | Field Surveyor |
|          18          | Jengo Tumaini  | +99820596465  |  None | 17 Port Harcourt Boulevard |     Akatsi    |  Kintampo | Field Surveyor |
|          20          |   Kunto Asha   | +99176320477  |  None |   30 Nyoka Achebe Street   |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+---------------+-------+----------------------------+---------------+-----------+----------------+

 CREATE EMAILS FOR ALL EMPLOYEES

In [40]:
%%sql 
# Update emails for employees so we can send them information

UPDATE md_water_services.employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov')
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

++
||
++
++

In [41]:
%%sql 
# Call up the employee table

SELECT 
    *
FROM  
    md_water_services.employee
LIMIT 15;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

15 rows affected.

+----------------------+----------------+---------------+-------------------------------+----------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  |  phone_number |             email             |          address           | province_name | town_name |    position    |
+----------------------+----------------+---------------+-------------------------------+----------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287  |   amara.jengo@ndogowater.gov  |  36 Pwani Mchangani Road   |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786  |   bello.azibo@ndogowater.gov  |   129 Ziwa La Kioo Road    |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041  |  bakari.iniko@ndogowater.gov  |   18 Mlima Tazama Avenue   |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900  | malachi.mavuso@ndogowater.gov |     100 Mogadishu Road     |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640  |  cheche.buhle@ndogowater.gov  |      1 Savanna Street      |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111  | zuriel.matembo@ndogowater.gov |  26 Bahari Ya Faraja Road  |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631  |  deka.osumare@ndogowater.gov  |    104 Kenyatta Street     |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240  | lalitha.kaburi@ndogowater.gov | 145 Sungura Amanpour Road  |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202  |   enitan.zuri@ndogowater.gov  |      117 Kampala Road      |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739  |    farai.nia@ndogowater.gov   | 33 Angélique Kidjo Avenue  |     Amanzi    |   Dahabu  | Field Surveyor |
|          12          |  Gamba Shani   | +99438283153  |   gamba.shani@ndogowater.gov  |   38 Mlima Tazama Avenue   |    Hawassa    |   Rural   | Field Surveyor |
|          14          |  Harith Nyota  | +99917230124  |  harith.nyota@ndogowater.gov  |      42 Malabo Drive       |    Kilimani   |   Rural   | Field Surveyor |
|          16          |  Isoke Amani   | +99313603182  |   isoke.amani@ndogowater.gov  |       55 Fennec Way        |     Sokoto    |   Rural   | Field Surveyor |
|          18          | Jengo Tumaini  | +99820596465  |  jengo.tumaini@ndogowater.gov | 17 Port Harcourt Boulevard |     Akatsi    |  Kintampo | Field Surveyor |
|          20          |   Kunto Asha   | +99176320477  |   kunto.asha@ndogowater.gov   |   30 Nyoka Achebe Street   |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+---------------+-------------------------------+----------------------------+---------------+-----------+----------------+

The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. However, when we use
the LENGTH(column) function, it returns 13 characters, indicating there's an extra character.
That's because there is a space at the end of the number! If you try to send an automated SMS to that number it will fail. The code below will fix this

In [94]:
%%sql

# TRIM the phone number column 

UPDATE employee
SET phone_number = TRIM(phone_number) 
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

56 rows affected.

++
||
++
++

# HONORING THE WORKERS 

Here, we want to honour and acknowledge the workers who spent time collecting the data, to make sure that we have enough information to solve the water crisis in Maji Ndogo

Before we dive into the analysis, let's have a look at where our employees live.

In [89]:
%%sql 
# NUMBER OF EMPLOYEES LIVING IN TOWNS

SELECT DISTINCT
    town_name,
    Count(employee_name) AS Number_of_employees
    
FROM  
    md_water_services.employee
GROUP BY town_name
LIMIT 15;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

9 rows affected.

+-----------+---------------------+
| town_name | Number_of_employees |
+-----------+---------------------+
|   Ilanga  |          3          |
|   Rural   |          29         |
|   Lusaka  |          4          |
|  Zanzibar |          4          |
|   Dahabu  |          6          |
|  Kintampo |          1          |
|   Harare  |          5          |
|  Yaounde  |          1          |
|   Serowe  |          3          |
+-----------+---------------------+

Note how many of our workers are living in smaller communities in the rural parts of Maji Ndogo.

Pres. Naledi congratulated the team for completing the survey, but we would not have this data were it not for our field workers. So let's gather
some data on their performance in this process, so we can thank those who really put all their effort in.


Pres. Naledi has asked we send out an email or message congratulating the top 3 field surveyors. So let's use the database to get the
employee_ids and use those to get the names, email and phone numbers of the three field surveyors with the most location visits.

In [92]:
%%sql 
# EMPLOYEE IDS WITH THE MOST NUMBER OF LOCATIONS VISITED 
SELECT DISTINCT
    assigned_employee_id,
    COUNT(Visit_count) AS Total_visit_count
FROM 
    md_water_services.visits
GROUP BY assigned_employee_id
ORDER BY COUNT(Visit_count) DESC
LIMIT 3;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

3 rows affected.

+----------------------+-------------------+
| assigned_employee_id | Total_visit_count |
+----------------------+-------------------+
|          1           |        3708       |
|          30          |        3676       |
|          34          |        3539       |
+----------------------+-------------------+

In [93]:
%%sql 
# FIND EMPLOYEE DETAILS WITH THE MOST NUMBER OF LOCATIONS VISITED 

SELECT
    *
FROM 
    employee
WHERE 
    assigned_employee_id
IN (1, 30, 34)


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

3 rows affected.

+----------------------+---------------+--------------+----------------------------+-----------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name | phone_number |           email            |        address        | province_name | town_name |    position    |
+----------------------+---------------+--------------+----------------------------+-----------------------+---------------+-----------+----------------+
|          1           |  Bello Azibo  | +99643864786 | bello.azibo@ndogowater.gov | 129 Ziwa La Kioo Road |    Kilimani   |   Rural   | Field Surveyor |
|          30          |   Pili Zola   | +99822478933 |  pili.zola@ndogowater.gov  |  46 Stone Town Street |    Hawassa    |  Zanzibar | Field Surveyor |
|          34          |   Rudo Imani  | +99046972648 | rudo.imani@ndogowater.gov  | 15 Lake Malawi Avenue |    Kilimani   |   Ilanga  | Field Surveyor |
+----------------------+---------------+--------------+----------------------------+-----------------------+---------------+-----------+----------------+

# ANALYSING LOCATIONS 

In [96]:
%%sql 

SELECT 
    province_name,
    town_name,
    location_type
FROM 
    location
LIMIT 15;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

15 rows affected.

+---------------+-----------+---------------+
| province_name | town_name | location_type |
+---------------+-----------+---------------+
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
|     Akatsi    |   Harare  |     Urban     |
+---------------+-----------+---------------+

Each record shows how many water sources and how many times they were visited by our employees. Lets explore the data to generate insights

In [125]:
%%sql 
# COUNT NUMBER  OF RECORDS PER TOWN
SELECT DISTINCT
    town_name,
     COUNT(*) AS Town_count
FROM 
    location
GROUP BY town_name
ORDER BY Town_count DESC
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+-----------+------------+
| town_name | Town_count |
+-----------+------------+
|   Rural   |   23740    |
|   Harare  |    1650    |
|   Amina   |    1090    |
|   Lusaka  |    1070    |
|   Mrembo  |    990     |
+-----------+------------+

In [128]:
%%sql
# COUNT NUMBER  OF RECORDS PER PROVINCE 
SELECT DISTINCT
    province_name,
     COUNT(*) AS Province_count
FROM 
    location
GROUP BY province_name
ORDER BY province_count DESC
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+---------------+----------------+
| province_name | Province_count |
+---------------+----------------+
|    Kilimani   |      9510      |
|     Akatsi    |      8940      |
|     Sokoto    |      8220      |
|     Amanzi    |      6950      |
|    Hawassa    |      6030      |
+---------------+----------------+

In [138]:
%%sql 
# COUNTS NUMBER OF RECORDS PER TOWN & PROVINCE
SELECT DISTINCT
    province_name,
    town_name,
     COUNT(town_name) AS records_per_town 
FROM 
    location
GROUP BY 
    province_name, 
    town_name 
ORDER BY 
    province_name,
    records_per_town DESC

LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+---------------+-----------+------------------+
| province_name | town_name | records_per_town |
+---------------+-----------+------------------+
|     Akatsi    |   Rural   |       6290       |
|     Akatsi    |   Lusaka  |       1070       |
|     Akatsi    |   Harare  |       800        |
|     Akatsi    |  Kintampo |       780        |
|     Amanzi    |   Rural   |       3100       |
+---------------+-----------+------------------+

Let's find out how many records we had for both urban and rural areas

In [141]:
%%sql
# NUMBER OF RECORDS FOR EACH LOCATION TYPE 

SELECT
    location_type,
    Count(*) AS num_of_records
FROM
    location
GROUP BY 
    location_type;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

2 rows affected.

+---------------+----------------+
| location_type | num_of_records |
+---------------+----------------+
|     Urban     |     15910      |
|     Rural     |     23740      |
+---------------+----------------+

# DIVING INTO SOURCES

Lets see how many people need access to water in Maji Ndogo (How many people did we survey in total?

In [145]:
%%sql 
# NUMBER OF PEOPLE SURVEYED
SELECT
    SUM(number_of_people_served)
FROM
    water_source
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+------------------------------+
| SUM(number_of_people_served) |
+------------------------------+
|           27628140           |
+------------------------------+

Lets see how many of each type of water source we have available in Maji Ndogo (How many wells, taps and rivers are there?)

In [147]:
%%sql 
# NUMBER OF EACH WATER SOURCE
SELECT
    type_of_water_source,
    COUNT(type_of_water_source) AS Total_num
FROM
    water_source
GROUP BY 
    type_of_water_source
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-----------+
| type_of_water_source | Total_num |
+----------------------+-----------+
|     tap_in_home      |    7265   |
|  tap_in_home_broken  |    5856   |
|         well         |   17383   |
|      shared_tap      |    5767   |
|        river         |    3379   |
+----------------------+-----------+

What is the average number of people that are served by each water source?

In [150]:
%%sql 
# AVERAGE NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    type_of_water_source,
    COUNT(type_of_water_source) AS Total_num,
    ROUND(AVG(number_of_people_served)) AS AVG_num_of_people_served
FROM
    water_source
GROUP BY 
    type_of_water_source
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+-----------+--------------------------+
| type_of_water_source | Total_num | AVG_num_of_people_served |
+----------------------+-----------+--------------------------+
|     tap_in_home      |    7265   |           644            |
|  tap_in_home_broken  |    5856   |           649            |
|         well         |   17383   |           279            |
|      shared_tap      |    5767   |           2071           |
|        river         |    3379   |           699            |
+----------------------+-----------+--------------------------+

These results are telling us that 644 people share a tap_in_home on average. Does that make sense?


No it doesn’t, right?
Remember I told you a few important things that apply to tap_in_home and broken_tap_in_home? The surveyors combined the data of many
households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of
6 people living in a home. So 6 people actually share 1 tap (not 644).
This means that 1 tap_in_home actually represents 644 ÷ 6 = ± 100 taps.

The total number of people served by each type of water source in total,

In [153]:
%%sql 
# TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    type_of_water_source,
    COUNT(type_of_water_source) AS Total_num_of_sources,
    SUM(number_of_people_served) AS Total_num_of_people_served
FROM
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    Total_num_of_people_served DESC
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+----------------------+----------------------------+
| type_of_water_source | Total_num_of_sources | Total_num_of_people_served |
+----------------------+----------------------+----------------------------+
|      shared_tap      |         5767         |          11945272          |
|         well         |        17383         |          4841724           |
|     tap_in_home      |         7265         |          4678880           |
|  tap_in_home_broken  |         5856         |          3799720           |
|        river         |         3379         |          2362544           |
+----------------------+----------------------+----------------------------+

In [161]:
%%sql 
# PERCENTAGE OF TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    type_of_water_source,
    COUNT(type_of_water_source) AS Total_num_of_sources,
    SUM(number_of_people_served) AS Total_num_of_people_served,
    ROUND(((SUM(number_of_people_served)/27628140)*100),0) AS Percentage_total_people_served_per_source
FROM
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    Total_num_of_people_served DESC
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

5 rows affected.

+----------------------+----------------------+----------------------------+-------------------------------------------+
| type_of_water_source | Total_num_of_sources | Total_num_of_people_served | Percentage_total_people_served_per_source |
+----------------------+----------------------+----------------------------+-------------------------------------------+
|      shared_tap      |         5767         |          11945272          |                     43                    |
|         well         |        17383         |          4841724           |                     18                    |
|     tap_in_home      |         7265         |          4678880           |                     17                    |
|  tap_in_home_broken  |         5856         |          3799720           |                     14                    |
|        river         |         3379         |          2362544           |                     9                     |
+----------------------+----------------------+----------------------------+-------------------------------------------+

# START OF A SOLUTION

At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision
how to do it. I think a simple approach is to fix the things that affect most people first. So let's write a query that ranks each type of source based
on how many people in total use it. RANK() should tell you we are going to need a window function to do this, so let's think through the problem.

In [166]:
%%sql 
# rank OF TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS Total_num_of_people_served,
    RANK() OVER(
    ORDER BY SUM(number_of_people_served) DESC) AS Rank_of_most_affected
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home'
GROUP BY 
    type_of_water_source
ORDER BY 
    Total_num_of_people_served DESC
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

4 rows affected.

+----------------------+----------------------------+-----------------------+
| type_of_water_source | Total_num_of_people_served | Rank_of_most_affected |
+----------------------+----------------------------+-----------------------+
|      shared_tap      |          11945272          |           1           |
|         well         |          4841724           |           2           |
|  tap_in_home_broken  |          3799720           |           3           |
|        river         |          2362544           |           4           |
+----------------------+----------------------------+-----------------------+

so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? We can use
the same logic; the most used sources should really be fixed first.

In [179]:
%%sql 
# rank OF TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    source_id,
    type_of_water_source,
    number_of_people_served,
    RANK() OVER(
    PARTITION BY type_of_water_source 
        ORDER BY number_of_people_served DESC) AS PRIORITY_RANK
    
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home'

LIMIT 50;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

50 rows affected.

+--------------+----------------------+-------------------------+---------------+
|  source_id   | type_of_water_source | number_of_people_served | PRIORITY_RANK |
+--------------+----------------------+-------------------------+---------------+
| SoRu34798224 |        river         |           998           |       1       |
| SoRu35837224 |        river         |           998           |       1       |
| SoRu36238224 |        river         |           998           |       1       |
| SoRu36791224 |        river         |           998           |       1       |
| SoRu36880224 |        river         |           998           |       1       |
| SoRu38142224 |        river         |           998           |       1       |
| SoRu37756224 |        river         |           998           |       1       |
| SoMa33775224 |        river         |           998           |       1       |
| KiRu30353224 |        river         |           998           |       1       |
| SoIl32972224 |        river         |           998           |       1       |
| KiZu31236224 |        river         |           998           |       1       |
| KiRu28591224 |        river         |           998           |       1       |
| KiMr25030224 |        river         |           998           |       1       |
| KiRu25603224 |        river         |           996           |       14      |
| KiRu26027224 |        river         |           996           |       14      |
| KiRu30532224 |        river         |           996           |       14      |
| SoRu36229224 |        river         |           996           |       14      |
| AkRu08053224 |        river         |           996           |       14      |
| AmAb09070224 |        river         |           996           |       14      |
| SoKo33179224 |        river         |           994           |       20      |
| HaRu20469224 |        river         |           994           |       20      |
| KiRu28404224 |        river         |           994           |       20      |
| KiRu27661224 |        river         |           994           |       20      |
| SoRu36506224 |        river         |           994           |       20      |
| SoRu37912224 |        river         |           994           |       20      |
| SoRu35192224 |        river         |           992           |       26      |
| SoRu36076224 |        river         |           992           |       26      |
| SoRu37032224 |        river         |           992           |       26      |
| SoMa33620224 |        river         |           992           |       26      |
| AkRu04142224 |        river         |           992           |       26      |
| SoRu38498224 |        river         |           992           |       26      |
| SoRu39029224 |        river         |           992           |       26      |
| AmRu13795224 |        river         |           992           |       26      |
| KiRu28002224 |        river         |           992           |       26      |
| KiRu27266224 |        river         |           992           |       26      |
| KiRu29844224 |        river         |           992           |       26      |
| SoRu37904224 |        river         |           990           |       37      |
| SoRu37158224 |        river         |           990           |       37      |
| SoRu37004224 |        river         |           990           |       37      |
| KiRu25651224 |        river         |           990           |       37      |
| SoIl32652224 |        river         |           990           |       37      |
| HaRu17065224 |        river         |           990           |       37      |
| AkRu08333224 |        river         |           990           |       37      |
| AkRu05337224 |        river         |           988           |       44      |
| AkRu06935224 |        river         |           988           |       44      |
| KiAm22552224 |        river         |           988           

Lets try using DENSE RANK and ROW NUMBER to see which ones would make more sense to the engineers 

In [180]:
%%sql 
# DENSE_RANK OF TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER(
    PARTITION BY type_of_water_source 
        ORDER BY number_of_people_served DESC) AS PRIORITY_RANK
    
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home'

LIMIT 50;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

50 rows affected.

+--------------+----------------------+-------------------------+---------------+
|  source_id   | type_of_water_source | number_of_people_served | PRIORITY_RANK |
+--------------+----------------------+-------------------------+---------------+
| SoRu34798224 |        river         |           998           |       1       |
| SoRu35837224 |        river         |           998           |       1       |
| SoRu36238224 |        river         |           998           |       1       |
| SoRu36791224 |        river         |           998           |       1       |
| SoRu36880224 |        river         |           998           |       1       |
| SoRu38142224 |        river         |           998           |       1       |
| SoRu37756224 |        river         |           998           |       1       |
| SoMa33775224 |        river         |           998           |       1       |
| KiRu30353224 |        river         |           998           |       1       |
| SoIl32972224 |        river         |           998           |       1       |
| KiZu31236224 |        river         |           998           |       1       |
| KiRu28591224 |        river         |           998           |       1       |
| KiMr25030224 |        river         |           998           |       1       |
| KiRu25603224 |        river         |           996           |       2       |
| KiRu26027224 |        river         |           996           |       2       |
| KiRu30532224 |        river         |           996           |       2       |
| SoRu36229224 |        river         |           996           |       2       |
| AkRu08053224 |        river         |           996           |       2       |
| AmAb09070224 |        river         |           996           |       2       |
| SoKo33179224 |        river         |           994           |       3       |
| HaRu20469224 |        river         |           994           |       3       |
| KiRu28404224 |        river         |           994           |       3       |
| KiRu27661224 |        river         |           994           |       3       |
| SoRu36506224 |        river         |           994           |       3       |
| SoRu37912224 |        river         |           994           |       3       |
| SoRu35192224 |        river         |           992           |       4       |
| SoRu36076224 |        river         |           992           |       4       |
| SoRu37032224 |        river         |           992           |       4       |
| SoMa33620224 |        river         |           992           |       4       |
| AkRu04142224 |        river         |           992           |       4       |
| SoRu38498224 |        river         |           992           |       4       |
| SoRu39029224 |        river         |           992           |       4       |
| AmRu13795224 |        river         |           992           |       4       |
| KiRu28002224 |        river         |           992           |       4       |
| KiRu27266224 |        river         |           992           |       4       |
| KiRu29844224 |        river         |           992           |       4       |
| SoRu37904224 |        river         |           990           |       5       |
| SoRu37158224 |        river         |           990           |       5       |
| SoRu37004224 |        river         |           990           |       5       |
| KiRu25651224 |        river         |           990           |       5       |
| SoIl32652224 |        river         |           990           |       5       |
| HaRu17065224 |        river         |           990           |       5       |
| AkRu08333224 |        river         |           990           |       5       |
| AkRu05337224 |        river         |           988           |       6       |
| AkRu06935224 |        river         |           988           |       6       |
| KiAm22552224 |        river         |           988           

In [181]:
%%sql 
# ROW_NUMBER rank OF TOTAL NUMBER OF PEOPLE SERVED BY EACH WATER SOURCE
SELECT
    source_id,
    type_of_water_source,
    number_of_people_served,
    ROW_NUMBER() OVER(
    PARTITION BY type_of_water_source 
        ORDER BY number_of_people_served) AS PRIORITY_RANK
    
FROM
    water_source
WHERE
    type_of_water_source != 'tap_in_home'

LIMIT 50;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

50 rows affected.

+--------------+----------------------+-------------------------+---------------+
|  source_id   | type_of_water_source | number_of_people_served | PRIORITY_RANK |
+--------------+----------------------+-------------------------+---------------+
| SoRu38500224 |        river         |           400           |       1       |
| SoRu39436224 |        river         |           400           |       2       |
| SoRu38759224 |        river         |           400           |       3       |
| SoRu38345224 |        river         |           400           |       4       |
| SoRu37948224 |        river         |           400           |       5       |
| SoRu37633224 |        river         |           400           |       6       |
| SoRu34808224 |        river         |           400           |       7       |
| KiRu27420224 |        river         |           400           |       8       |
| AkKi01493224 |        river         |           400           |       9       |
| SoMa33669224 |        river         |           400           |       10      |
| SoMa33948224 |        river         |           400           |       11      |
| SoKo33424224 |        river         |           400           |       12      |
| KiMr24417224 |        river         |           400           |       13      |
| KiHa23390224 |        river         |           400           |       14      |
| AmRu14978224 |        river         |           400           |       15      |
| HaDj16848224 |        river         |           400           |       16      |
| HaAm15940224 |        river         |           402           |       17      |
| SoRu34889224 |        river         |           402           |       18      |
| SoRu37206224 |        river         |           402           |       19      |
| SoRu35540224 |        river         |           402           |       20      |
| SoRu34900224 |        river         |           402           |       21      |
| SoRu35362224 |        river         |           402           |       22      |
| KiRu29598224 |        river         |           402           |       23      |
| KiRu27306224 |        river         |           402           |       24      |
| KiMr24831224 |        river         |           402           |       25      |
| KiIs23538224 |        river         |           402           |       26      |
| KiIs23568224 |        river         |           402           |       27      |
| KiIs24082224 |        river         |           402           |       28      |
| AmAm09712224 |        river         |           402           |       29      |
| SoIl32582224 |        river         |           402           |       30      |
| SoIl32326224 |        river         |           402           |       31      |
| KiRu30631224 |        river         |           402           |       32      |
| KiRu30356224 |        river         |           404           |       33      |
| SoBa31550224 |        river         |           404           |       34      |
| KiRu30113224 |        river         |           404           |       35      |
| SoRu34764224 |        river         |           404           |       36      |
| SoMa34500224 |        river         |           404           |       37      |
| SoRu37294224 |        river         |           404           |       38      |
| SoRu34944224 |        river         |           404           |       39      |
| AkRu05536224 |        river         |           404           |       40      |
| SoRu38482224 |        river         |           404           |       41      |
| KiRu27249224 |        river         |           404           |       42      |
| KiRu29975224 |        river         |           404           |       43      |
| KiRu28466224 |        river         |           404           |       44      |
| KiRu26987224 |        river         |           404           |       45      |
| AkRu06038224 |        river         |           406           

# ANALYSING QUEUES

The visits table documented all of the visits our field surveyors made to each location. For most sources, one visit was enough, but if there were
queues, they visited the location a couple of times to get a good idea of the time it took for people to queue for water. So we have the time that
they collected the data, how many times the site was visited, and how long people had to queue for water.


How long did the survey take?

In [192]:
%%sql 
# CALCULATE HOW LONG THE SURVEY TOOK
SELECT 
    SUBSTRING(MIN(time_of_record),1,10) AS Survey_start_date,
    SUBSTRING(MAX(time_of_record),1,10) AS Survey_end_date,
    DATEDIFF(SUBSTRING(MAX(time_of_record),1,10), SUBSTRING(MIN(time_of_record),1,10)) AS Total_survey_time_in_days
    
FROM
    visits;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+-------------------+-----------------+---------------------------+
| Survey_start_date | Survey_end_date | Total_survey_time_in_days |
+-------------------+-----------------+---------------------------+
|     2021-01-01    |    2023-07-14   |            924            |
+-------------------+-----------------+---------------------------+

Let's see how long people have to queue on average in Maji Ndogo. Keep in mind that many sources like taps_in_home have no queues. These
are just recorded as 0 in the time_in_queue column, so when we calculate averages, we need to exclude those rows.

In [199]:
%%sql 
# What is the average total queue time for water if they dont have taps in their homes in minutes?

SELECT 
    AVG(NULLIF(time_in_queue,0)) AS Average_queue_time
FROM 
    visits

    
    
LIMIT 5;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

1 rows affected.

+--------------------+
| Average_queue_time |
+--------------------+
|      123.2574      |
+--------------------+

So let's look at the queue times aggregated across the different days of the week.

In [215]:
%%sql
# What is the average queue time on different days?

SELECT DISTINCT
    DAYNAME(time_of_record) AS Day_of_week,
    ROUND(AVG(NULLIF(time_in_queue,0)),0) AS Average_queue_time_in_minutes
FROM
    Visits
GROUP BY 
	DAYNAME(time_of_record)
;


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

7 rows affected.

+-------------+-------------------------------+
| Day_of_week | Average_queue_time_in_minutes |
+-------------+-------------------------------+
|    Friday   |              120              |
|   Saturday  |              246              |
|    Sunday   |               82              |
|    Monday   |              137              |
|   Tuesday   |              108              |
|  Wednesday  |               97              |
|   Thursday  |              105              |
+-------------+-------------------------------+

We can also look at what time during the day people collect water.

In [223]:
%%sql 
# what time during the day people collect water.

SELECT DISTINCT
    TIME_FORMAT(Time(time_of_record), '%H:00') AS Hour_of_day,
    ROUND(AVG(NULLIF(time_in_queue,0)),0) AS Average_queue_time_in_minutes
FROM
    Visits
GROUP BY 
	TIME_FORMAT(Time(time_of_record), '%H:00')
ORDER BY 
    Hour_of_day 
;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

+-------------+-------------------------------+
| Hour_of_day | Average_queue_time_in_minutes |
+-------------+-------------------------------+
|    06:00    |              149              |
|    07:00    |              149              |
|    08:00    |              149              |
|    09:00    |              118              |
|    10:00    |              114              |
|    11:00    |              111              |
|    12:00    |              112              |
|    13:00    |              115              |
|    14:00    |              114              |
|    15:00    |              114              |
|    16:00    |              114              |
|    17:00    |              149              |
|    18:00    |              147              |
|    19:00    |              168              |
+-------------+-------------------------------+

Here we calculate the average time in queue for hours of the day for each day of the week 

In [236]:
%%sql
SELECT
TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
-- Sunday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Sunday,
    
-- Monday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Monday,
      
-- Tuesday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Tuesday,
    
-- Wednesday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Wednesday,
    
-- Thursday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Thursday,
    
-- Friday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Friday,
    
-- Saturday
    ROUND(AVG(
    CASE
    WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
    ELSE NULL
    END
    ),0) AS Sarurday
    
    
    
FROM
visits
WHERE
time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
hour_of_day
ORDER BY
hour_of_day;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Sarurday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
|    16:00    |   83   |  127   |    99   |     90    |    99    |  109   |   251    |
|    17:00    |   79   |  181   |   135   |    121    |   129    |  151   |   251    |
|    18:00    |   80   |  174   |   122   |    113    |   132    |  158   |   240    |
|    19:00    |  127   |  159   |   145   |    176    |   137    |  103   |   282    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+

Some insight: 
1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing
for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. The people of Maji Ndogo prioritise family and religion, so Sundays are spent
with family and friends.

Other Insights
1. Most water sources are rural.
2. 43% of our people are using shared taps. 2000 people often share one tap.
3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes,
pumps, and reservoirs.
4. 18% of our people are using wells of which, but within that, only 28% are clean..
5. Our citizens often face long wait times for water, averaging more than 120 minutes.
6. In terms of queues:
- Queues are very long on Saturdays.
- Queues are longer in the mornings and evenings.
- Wednesdays and Sundays have the shortest queues.